### PR TITLE
Add sync method to the Secrets client

### LINKED ISSUE #6198: Implement "sync" Secret APIs

### PR DESCRIPTION
**Pull request by AlexGhiondea**
This also switches over to the IAsyncEnumerable from Azure.Core.

Fixes #6198

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 1]
```diff
@@ -54,33 +54,41 @@ public virtual async Task<Response<Secret>> GetAsync(string name, string version
...
 
-        public virtual AsyncEnumerator<SecretBase> GetAllVersionsAsync(string name, CancellationToken cancellationToken = default)
+        public virtual IAsyncEnumerable<Response<SecretBase>> GetAllVersionsAsync(string name, CancellationToken cancellationToken = default)
         {
             if (string.IsNullOrEmpty(name)) throw new ArgumentNullException(nameof(name));
 
             Uri firstPageUri = new Uri(_vaultUri, $"{SecretsPath}{name}/versions");
 
-            return new AsyncEnumerator<SecretBase>(firstPageUri, () => new SecretBase(), this.GetPageAsync<SecretBase>, cancellationToken);
+            return PageResponseEnumerator.CreateAsyncEnumerable(nextLink => this.GetPageAsync(firstPageUri, ()=> new SecretBase(), cancellationToken));
```
Comment (maririos) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: you shouldn't need the `this.`

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 2]
```diff
@@ -258,16 +329,49 @@ private async Task<Response> SendRequestAsync(Request request, CancellationToken
...
-            where T : Model
+        private async Task<PageResponse<T>> GetPageAsync<T>(Uri pageUri, Func<T> itemFactory, CancellationToken cancellationToken)
+                where T : Model
         {
             using (Request request = CreateRequest(HttpPipelineMethod.Get, pageUri))
             {
                 Response response = await SendRequestAsync(request, cancellationToken);
 
-                return this.CreateResponse(response, new Page<T>(itemFactory));
+                // read the respose
```
Comment (maririos) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: We should check the status of the response too. See as example: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/appconfiguration/Azure.ApplicationModel.Configuration/src/ConfigurationClient.cs#L517

Comment (AlexGhiondea) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: Done. There is code in `SendRequestAsync` that checks the status before returning the response.

**Review by maririos — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 3]
```diff
@@ -104,14 +116,20 @@ public virtual async Task<Response<Secret>> SetAsync(Secret secret, Cancellation
...
         }
 
         public virtual Response<Secret> Set(Secret secret, CancellationToken cancellationToken = default)
         {
-            throw new NotImplementedException();
+            if (secret == null) throw new ArgumentNullException(nameof(secret));
+
+            if (secret.Name == null) throw new ArgumentNullException($"{nameof(secret)}.{nameof(secret.Name)}");
+
+            if (secret.Value == null) throw new ArgumentNullException($"{nameof(secret)}.{nameof(secret.Value)}"); ;
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]:
```suggestion
            if (secret.Value == null) throw new ArgumentNullException($"{nameof(secret)}.{nameof(secret.Value)}");
```

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 4]
```diff
@@ -245,11 +245,11 @@ public virtual async Task<Response<SecretBase>> RestoreAsync(byte[] backup, Canc
             return await SendRequestAsync(HttpPipelineMethod.Post, new VaultBackup { Value = backup }, () => new SecretBase(), cancellationToken, SecretsPath, "/restore");
         }
 
-        public virtual Response<Secret> Restore(byte[] backup, CancellationToken cancellationToken = default)
+        public virtual Response<SecretBase> Restore(byte[] backup, CancellationToken cancellationToken = default)
```
Comment (maririos) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: I see you are applying some of the changes that @schaabs  has in PR https://github.com/Azure/azure-sdk-for-net/pull/6436/files but not all and that might be why you are adding `/` and stuff.
Is the idea to merge this one and then update the other PR? or are you going to take all of the changes in that PR here?

Comment (AlexGhiondea) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: I will get the changes from both into this PR :).

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 5]
```diff
@@ -259,25 +329,70 @@ private async Task<Response> SendRequestAsync(Request request, CancellationToken
...
+
+            using (Request request = CreateRequest(HttpPipelineMethod.Get, firstPageUri, addAPIVersion: false))
+            {
+                Response response = SendRequest(request, cancellationToken);
+
+                // read the respose
+                Page<T> responseAsPage = new Page<T>(itemFactory);
+                responseAsPage.Deserialize(response.ContentStream);
+
+                // convert from the Page<T> to PageResponse<T>
```
Comment (pakrym) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: We need to get rid of `Page<T>` completely.

Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: Hm, this ties into the question of having a `Model` base type at all. Let's skip for this PR.

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs [hunk 6]
```diff
@@ -352,7 +352,7 @@ private async Task<PageResponse<T>> GetPageAsync<T>(Uri firstPageUri, string nex
                 firstPageUri = new Uri(nextLink);
             }
 
-            using (Request request = CreateRequest(HttpPipelineMethod.Get, firstPageUri, addAPIVersion: false))
```
Comment (AlexGhiondea) [inline sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: We no longer need to be clever with when we add the API version?

Comment (pakrym) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: We never set this parameter to anything but `false`

**Review by pakrym — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs (hunk 5 above)
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Secrets/src/SecretClient.cs]: Issue https://github.com/Azure/azure-sdk-for-net/issues/6477 created to address that change. I added both `Model` and `Page`

**Pull request merged**